TSLint → ESLint

stevenvachon opened the pull request:
<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/master/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the master branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

-->

**What this PR does / why we need it**:
Migrates from the now deprecated [tslint](https://npmjs.com/tslint) to [eslint](https://npmjs.com/eslint) (and the necessary plugins for TypeScript).  Some basic formatting rules were moved out of "lint" configs and into `.editorconfig` for reduced redundancy.

With that, some supportive changes were necessary (#21441, #21586) ~~and they should be merged before this~~.

**Which issue(s) this PR fixes**:
Fixes #16053, #19345
Indirectly fixes #20430

**Special notes for your reviewer**:
One liberty taken was to separate our [eslintrc](https://eslint.org/docs/user-guide/configuring) and [tsconfig](https://www.typescriptlang.org/docs/handbook/tsconfig-json.html) to their own packages: [@grafana/eslint-config](https://npmjs.com/@grafana/eslint-config) and [@grafana/tsconfig](https://npmjs.com/@grafana/tsconfig). Instead of parent referencing (`../`) directories in many places, a package name is used and fewer files overall are necessary. For me personally, this simplified my mental graph of the codebase that I am still learning, which I think will translate to colleagues and third-party plugin developers.

stevenvachon committed:
Alphabetized tslint and tsconfig files

stevenvachon committed:
Replaced TSLint with ESLint

stevenvachon committed:
Switched to @grafana/tsconfig

… and:
* de-duped options
* removed options with default values

claassistantio commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/grafana?pullRequest=21006) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/grafana/grafana?pullRequest=21006) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/grafana?pullRequest=21006) it.</sub>

ryantxu commented:
> @torkelo @dprokop -- I was told that you might be concerned with this change.

😄 I said they will have opinions!

My biggest concern is with polluting the package names with many small packages.  That may be a non-concern or something we can handle with folders or prefixes.  This currently proposes:
![image](https://user-images.githubusercontent.com/705951/70562487-57516e00-1b41-11ea-9665-6e5afe801271.png)

Maybe changing the folder names to: `config-tslint` and `config-tsconfig`?  would group them better?

torkelo commented:
Cant the tsconfig and eslint config live in toolkit package or do they have to be in their own packages?

stevenvachon commented:
I don't think that it makes sense to depend on the toolkit package for lints to work correctly within other unrelated packages. The purpose of a monorepo is to easily make changes across multiple—but still isolated—packages, not to treat them as a single project.

torkelo commented:
But the toolkit package is designed to contain some underlying building blocks and build scripts, shared config etc, exactly things tsconfig and lint config so creating a separate packages with a single file seems too granular, but maybe warranted / good idea, just will increase time running workspace commands

dprokop commented:
> But the toolkit package is designed to contain some underlying building blocks and build scripts, shared config etc, exactly things tsconfig and lint config so creating a separate packages with a single file seems too granular, but maybe warranted / good idea, just will increase time running workspace commands

@torkelo as far as I see this PR does not introduce changes to the toolkit's internal tsconfig (common config for plugin developers) but only for how the packages are using the shared config. So grafana-toolkit still uses tslint as a driver for plugin linting. But I don't think changing this is a goal of this PR, isn't it @stevenvachon? I think having the tslint/tsconfig packages would be useful for plugin devs so that they don't have to extend tsconfig in a way they need to now:

```
  "extends": "./node_modules/@grafana/toolkit/src/config/tsconfig.plugin.json",
```
would become
```
  "extends": "@grafana/tsconfig",
```

I think the shared packages are not necessarily a bad idea- it's a common practice to have configs available as packages. But those packages will have to be also published to npm for the plugin developers to take advantage from. 

@stevenvachon do I understand correctly that this PR replaces TSLint with ESLint in the core Grafana too?

torkelo commented on package.json:
```diff
@@ -157,7 +162,7 @@
     "build": "grunt build",
     "test": "grunt test",
-    "tslint": "tslint -c tslint.json --project tsconfig.json",
+    "tslint": "eslint public/app public/e2e-test public/test public/vendor --ext=.js,.ts,.tsx",
```
don't think we need to lint public/vendor, old vendored libs that will have hundreds of errors

torkelo commented on packages/grafana-eslint-config/test.js:
```diff
@@ -0,0 +1,7 @@
+'use strict';
+const {expect} = require('chai');
```
This looks like prettier is not formatting, need to update the root package.json prettier commands to also run on js files :)

stevenvachon replied:
Yeah, that's just something temporary that I'd created quickly to outline my plans.

torkelo reviewed (commented):
yea, the extends become a bit shorter if they are at the root of their own package. 

The CI build error looks strange, yarn seems not to be able to link in the package

stevenvachon commented:
> So grafana-toolkit still uses tslint as a driver for plugin linting.

No, grafana-toolkit will soon use ESLint as well. TSLint is deprecated.

> @stevenvachon do I understand correctly that this PR replaces TSLint with ESLint in the core Grafana too?

Correct.

stevenvachon committed:
TSLint disables → ESLint disables

… also JSHint removals, which haven’t had an affect since it was replaced with TSLint.

torkelo commented:
once this is merged, would be nice to get it to output junit report so circle ci can report eslint error in a formatted way at the top: 
https://www.viget.com/articles/using-junit-on-circleci-2-0-with-jest-and-eslint/ 

i recently did this for frontend tests, and it greatly improves finding failures in CI results:
https://github.com/grafana/grafana/pull/21084

ryantxu commented:
Looking at this a bit more...  I don't like that we would be adding two packages `grafana-eslint-config` and `grafana-tsconfig` to the main grafana build process.

It also seems a little weird to have the config files version locked to the grafana version.

Maybe a better option is to create two new repos:
```
eslint-config-grafana
tsconfig-grafana
```
And publish to npm with that name (not @grafana/...)  This would follow what airbnb does

@dprokop thoughts?

torkelo commented:
Maybe these two packages can be seperate npm packages @grafana/es-lint-config & @grafana/tsconfig but maybe we can put them in a separate repo (grafana/npm-packages)? we wont be able to reuse our npm publish scripts but that could be minimal?

But then again, how much will adding these two projects add to the build process? They wont require any building really, only publish? 

@dprokop what do you think?

dprokop commented:
Regarding eslint config. There is an official ESLint recommendation about the naming that can be found here: https://eslint.org/docs/developer-guide/shareable-configs#creating-a-shareable-config

**Naming**
In summary the package name should be prefixed with `eslint-config-` which in our case the name shoud be either `eslint-config-grafana` or `@grafana/eslint-config-grafana`. For the matter of consistency and organization scoping I would go with `@grafana/eslint-config-grafana`, and the same with tsconfig configuration: `@grafana/tsconfig-grafana`. 

**Publishing**
This should require minor changes so that eslint/tsconfig packages are published only when the actual package changes. Currently ALL of the @grafana/* packages are published (canary) when there are changes detected in ANY of the packages. I guess the config should be released independently of that rule. Also, there will be no build step for these two packages so publishing them in the main build pipeline should not impact us much - especially taking the config changes rate into account.

**Versioning**
Not sure what's the best approach here. Agree with @ryantxu that having the config version locked to Grafana version feels weird. Some thoughts on this @torkelo @stevenvachon ?

ryantxu commented:
+1 for:
```
@grafana/eslint-config-grafana
@grafana/tsconfig-grafana
```

For versioning, I would put each in its own repository, I suggest matching the names above.

torkelo commented:
I think they can be in this repo,I talked to dominik and we do not need to publish these packages on every build , so it will not slow down the build.

And having to duplicate our npm package publish scripts and pipeline seems bad

ryantxu commented:
My argument for having them in a different repo is that the deployment and versions are totally different -- I think it should be explicit and manual in a way the packages/* are not.

If we do keep it in the same repo, maybe a different folder?

dprokop commented:
I think if we are deciding to keep these two in the same repo we should keep them in the same directory as other packages. These packages will not be automatically released with our pipeline unless directly specified in the release scripts, so we should be fine here. The one thing that might be problematic is the `packages:prepare` task that updates the cross-package versions to the lerna.json specified one, which means that config package versions will be updated as well - not sure what is the workaround for that apart from manually modyfing the versions back to the correct one in the script.

Having these two packages in another directory is not so good as we will introduce inconsistency in package.json and lerna.json about which packages(versioning) are managed by what pipeline (core packages versioning managed by lerna, config packages managed manually). Also we would need to have two directories for workspaces which feels weird.

@torkelo we will have to have separate publishing scripts for config packages anyways as core packages are managed by Lerna (for config packages it's actually not a script, but just an npm script running `npm publish`).

Taking into account the frequency of changes in the config packages, different versioning schema and simplified publishing I think we can have them as separate repo(s).

ryantxu commented on packages/grafana-toolkit/src/cli/tasks/plugin.build.ts:
```diff
@@ -1,173 +1,97 @@
-  const globPattern = path.resolve(process.cwd(), 'src/**/*.+(ts|tsx)');
-  return glob.sync(globPattern);
-};
+const getTypescriptSources = () => globby(resolvePath(process.cwd(), 'src/**/*.+(ts|tsx)'));
```
are these `globby` changes related to TSLint -> ESLint?  

I like the change, but maybe we can pull that into a different PR and get it committed quickly?

stevenvachon replied:
They'll be split out to a separate "cleanup" PR when all the work is done.  It's currently all part of a related train of thought.

torkelo commented:
> @torkelo we will have to have separate publishing scripts for config packages anyways as core packages are managed by Lerna (for config packages it's actually not a script, but just an npm script running npm publish). 

That settles it then, seperate repo!

ryantxu commented:
I just created:
https://github.com/grafana/eslint-config-grafana
https://github.com/grafana/tsconfig-grafana

and copied the files from this PR to each of them.  @stevenvachon -- can you update the structure as needed.  (not sure when it is easiest to do... and to start publishing to npm)

stevenvachon committed:
Optimized editorconfig & prettier config files

… to reduce redundancy

stevenvachon committed:
Compliances for ESLint, Prettier and TypeScript

stevenvachon committed:
Optimized tsconfig files

ryantxu commented on packages/grafana-data/package.json:
```diff
@@ -14,13 +14,15 @@
   },
   "devDependencies": {
+    "@grafana/eslint-config": "^1.0.0-pre",
+    "@grafana/tsconfig": "^1.0.0-pre",
```
Should this be: [@grafana/tsconfig-grafana](https://github.com/grafana/tsconfig-grafana/blob/master/package.json#L4)

stevenvachon replied:
Yes, I haven't renamed/moved them yet.

ryantxu commented on packages/grafana-data/package.json:
```diff
@@ -14,13 +14,15 @@
     "build": "grafana-toolkit package:build --scope=data"
   },
   "devDependencies": {
+    "@grafana/eslint-config": "^1.0.0-pre",
```
Should this be [@grafana/eslint-config-grafana](https://github.com/grafana/eslint-config-grafana/blob/master/package.json#L4)

stevenvachon replied:
Yes, I haven't renamed/moved them yet.

stevenvachon commented:
Hmm:
> npm scoped modules are also supported, by naming or prefixing the module with `@scope/eslint-config`, such as `@scope/eslint-config` **or** `@scope/eslint-config-myconfig`.

I think `@scope/eslint-config` is less redundant.

torkelo commented:
yea, @scope/eslint-config looks fine

dprokop commented:
@stevenvachon +1 for @scope/eslint-config

stevenvachon committed:
Fixed nasty issue with types for nested slate-react

stevenvachon commented:
The following scripts are working and passing:
* `yarn build`
* `yarn lint`
* `yarn typecheck`
* `yarn packages:build`
* `yarn packages:lint`
* `yarn packages:typecheck`

The following fails miserably:
* `yarn test`

If anyone (@dprokop, @ryantxu, @torkelo) can take a look to help solve this, would likely expedite completion.

ryantxu commented:
for the "StringMap expected string key" maybe:
https://medium.com/@gavin.blog/how-to-solve-ng-annotate-stringmap-expected-string-key-error-f0711c7c84db

stevenvachon committed:
Updated lockfile

dprokop commented on package.json:
```diff
@@ -59,6 +59,8 @@
     "@types/slate-react": "0.22.5",
     "@types/testing-library__react-hooks": "^3.1.0",
     "@types/tinycolor2": "1.4.2",
+    "@typescript-eslint/eslint-plugin": "^2.18.0",
```
```suggestion
    "@typescript-eslint/eslint-plugin": "2.18.0",
```

stevenvachon replied:
Why when we have a lockfile?

dprokop replied:
We had couple of issues in the past when regenerated yarn.lock(from scratch) was referencing versions which were causing troubles.

dprokop commented on package.json:
```diff
@@ -59,6 +59,8 @@
     "@types/testing-library__react-hooks": "^3.1.0",
     "@types/tinycolor2": "1.4.2",
+    "@typescript-eslint/eslint-plugin": "^2.18.0",
+    "@typescript-eslint/parser": "^2.18.0",
```
```suggestion
    "@typescript-eslint/parser": "2.18.0",
```

dprokop commented on package.json:
```diff
@@ -73,6 +75,12 @@
     "enzyme-to-json": "3.3.5",
     "es6-promise": "3.3.1",
     "es6-shim": "0.35.5",
+    "eslint": "^6.8.0",
```
Please fix the version of dependencies

dprokop commented on public/app/core/services/ng_react.ts:
```diff
@@ -54,7 +54,7 @@ function applied(fn: any, scope: any) {
     return fn;
   }
-  //tslint:disable-next-line:only-arrow-functions
+  // this had the equivalent of `eslint-disable-next-line prefer-arrow/prefer-arrow-functions`
```
Not sure what you mean by this comment?

stevenvachon replied:
Some tslint rules are slightly different in eslint (and typescript-eslint), which would've resulted in many source code file changes. They'll be handled soon in another PR, but I wanted to keep a record of the previous state without retaining any tslint comments.

dprokop replied:
gottcha

dprokop reviewed (commented):
This is looking good, great stuff @stevenvachon . There is one thing that I find missing. In the packages README I'd like to see a guide on how and where from to publish the new configs. This should also be documented in config's READMEs. 

This also introduces a *breaking change* in the grafana/toolkit. I think we need to add some kind of guide on how to migrate. Maybe try automating the tslint/tsconfig -> eslint migration in toolkit lint task?

ryantxu reviewed (approved):
LGTM -- lets fix any migration guides in a follow on issue
